File: generated_code/cpp/solution_671_2.cpp
#include <algorithm>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        // Base case: if the root is null or a leaf node, there's no second minimum.
        if (!root || !root->left) return -1;
        
        // Since root.val is the minimum of its children, the smallest value in the 
        // whole tree is always root->val.
        int left_val = root->left->val;
        int right_val = root->right->val;
        
        // If the left child's value equals the root's value, the second minimum 
        // could be located deeper in the left subtree. We recursively search for it.
        if (left_val == root->val) {
            left_val = findSecondMinimumValue(root->left);
        }
        
        // Similarly, if the right child's value equals the root's value, the second 
        // minimum could be located deeper in the right subtree.
        if (right_val == root->val) {
            right_val = findSecondMinimumValue(root->right);
        }
        
        // If both subtrees returned a valid second minimum, the overall second 
        // minimum for the current root is the smaller of the two.
        if (left_val != -1 && right_val != -1) {
            return std::min(left_val, right_val);
        }
        
        // If only one subtree returned a valid second minimum, return that one.
        // If both returned -1, then -1 is returned.
        return left_val != -1 ? left_val : right_val;
    }
};

File: generated_code/python/solution_671_2.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def findSecondMinimumValue(self, root: Optional[TreeNode]) -> int:
        """
        Algorithm: DFS with Pruning
        Time Complexity: O(N) where N is the number of nodes in the tree.
        Space Complexity: O(H) where H is the height of the tree for the recursion stack.
        
        The root node always holds the minimum value of the entire tree because
        root.val = min(root.left.val, root.right.val) holds everywhere.
        We can find the second minimum by scanning the tree. If a node's value is
        strictly greater than the root's value, it is a candidate for the second minimum,
        and we don't need to look into its subtrees (since all values below it will be >= it).
        """
        if not root:
            return -1
            
        # The absolute minimum value in the tree is always at the root.
        min_val = root.val
        self.second_min = float('inf')
        
        def dfs(node: Optional[TreeNode]) -> None:
            if not node:
                return
            
            # If the current node's value is strictly greater than the absolute minimum,
            # it's a potential second minimum candidate.
            if min_val < node.val < self.second_min:
                self.second_min = node.val
                
            # If a node's value equals the root value, its children might contain
            # the second minimum, so we continue exploring.
            # If a node's value is already greater than min_val, any value in its 
            # subtree will be >= node.val, which cannot be smaller than our current 
            # node.val candidate. Thus, we can safely prune and stop searching its subtree.
            if node.val == min_val:
                dfs(node.left)
                dfs(node.right)
                
        dfs(root)
        
        # If second_min was updated, return it; otherwise, return -1.
        return self.second_min if self.second_min != float('inf') else -1